src/uct/tcp/tcp_ep.c

```diff
@@ -329,139 +367,265 @@ void uct_tcp_ep_set_failed(uct_tcp_ep_t *ep)
-                                                   ((struct sockaddr*)params->dev_addr);
+    ucs_assert(ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP);
 
-    conn_sn = uct_tcp_cm_get_conn_sn(iface, &dest_addr);
+    addr->conn_sn = ep->conn_sn;
+    return uct_iface_get_address(tl_ep->iface,
+                                 (uct_iface_addr_t*)&addr->iface_addr);
+}
 
-    if (uct_tcp_iface_is_self_addr(iface, &dest_addr)) {
-        goto out_create_ep;
+ucs_status_t uct_tcp_ep_connect_to_ep(uct_ep_h tl_ep,
+                                      const uct_device_addr_t *dev_addr,
+                                      const uct_ep_addr_t *ep_addr)
+{
+    uct_tcp_ep_t *ep        = ucs_derived_of(tl_ep, uct_tcp_ep_t);
+    uct_tcp_iface_t *iface  = ucs_derived_of(ep->super.super.iface,
+                                             uct_tcp_iface_t);
+    uct_tcp_ep_addr_t *addr = (uct_tcp_ep_addr_t*)ep_addr;
+    uct_tcp_ep_t *peer_ep;
+    int connect_to_self;
+
+    ucs_assert(ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP);
+
+    uct_tcp_ep_set_dest_addr(dev_addr, (uct_iface_addr_t*)&addr->iface_addr,
+                             &ep->peer_addr);
+
+    connect_to_self = uct_tcp_ep_is_self(ep);
+    if (!uct_tcp_cm_ep_accept_conn(ep) || connect_to_self) {
+        if (!connect_to_self) {
+            /* EP that are created as CONNECT_TO_EP has to be full-duplex, set RX
+             * capability as well as TX (that's set in uct_tcp_ep_connect()) */
+            uct_tcp_ep_add_ctx_cap(ep, UCT_TCP_EP_FLAG_CTX_TYPE_RX);
+        }
+        return uct_tcp_ep_connect(ep);
     }
 
-    ep = uct_tcp_cm_get_ep(iface, &dest_addr, conn_sn,
-                           UCT_TCP_EP_FLAG_CTX_TYPE_RX);
-    if (ep == NULL) {
-        goto out_create_ep;
-    }
+    /* Use remote peer connection sequence number value, since the EP has to
+     * accept the conenction */
+    ep->conn_sn = addr->conn_sn;
 
-    ucs_assert(!(ep->flags & UCT_TCP_EP_FLAG_CTX_TYPE_TX));
-    /* Found EP with RX ctx, try to send the connection request
-     * to the remote peer, if it successful - assign TX to this EP
-     * and return the EP to the user */
-    status = uct_tcp_cm_send_event(ep, UCT_TCP_CM_CONN_REQ, 0);
-    if (status != UCS_OK) {
-        uct_tcp_ep_destroy_internal(&ep->super.super);
-        ep = NULL;
+    peer_ep = uct_tcp_cm_get_conn_to_ep(iface, &ep->peer_addr, ep->conn_sn,
+                                        UCT_TCP_EP_FLAG_CTX_TYPE_RX);
+    if (peer_ep != NULL) {
+        ucs_assert(!(peer_ep->flags & UCT_TCP_EP_FLAG_CTX_TYPE_TX) &&
+                   (peer_ep->fd != -1));
+        ep->conn_retries++;
+        uct_tcp_ep_move_ctx_cap(peer_ep, ep, UCT_TCP_EP_FLAG_CTX_TYPE_RX);
+        uct_tcp_ep_add_ctx_cap(ep, UCT_TCP_EP_FLAG_CTX_TYPE_TX);
+        uct_tcp_ep_replace_ep(ep, peer_ep);
+        peer_ep = NULL;
+
+        uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_CONNECTING);
```

Comment:
why need to go thru both states in a row?

Reply:
removed, but updated `uct_tcp_cm_change_conn_state()` to support changing from CLOSED to CONNECTED